src/uwuw/uwuw_preprocessor.cpp

```diff
@@ -0,0 +1,545 @@
+
+  // load the material objects
+  // load the dag file
+  moab::ErrorCode rval = DAG->load_file(dagmc_filename.c_str());
+
+  // do the minimal DAGMC initialisation
+  rval = DAG->setup_impl_compl();
+  rval = DAG->setup_indices();
+
+  // make a new dagmcmetadata class
+  dmd = new dagmcMetaData(DAG);
+  dmd->load_property_data();
+  // set the output filename
+  output_filename = output_file;
+
+  // set the verbosity
+  verbose = verbosity;
+  fatal = fatal_errors;
+}
+
+// destructor
+uwuw_preprocessor::~uwuw_preprocessor() {
+}
+
+// write the new material library
+void uwuw_preprocessor::write_uwuw_materials() {
+  std::map<std::string, pyne::Material> :: iterator it;
+
+  // loop over the processed material library and write each one to the file
+  for (it = uwuw_material_library.begin() ; it != uwuw_material_library.end() ; ++it) {
+    // the current material
+    pyne::Material mat = it->second;
+    // write the material to the file
+    if (verbose) {
+      std::cout << "writing material, " << mat.metadata["name"].asString();
+      std::cout << "writing material, " << mat.metadata["fluka_name"].asString();
+      std::cout << " to file " << output_filename << std::endl;
+    }
+    // write the uwuw materials to the geometry
+    mat.write_hdf5(output_filename, "/materials");
+  }
+
+  return;
+}
+
+// write the new material library
+void uwuw_preprocessor::write_uwuw_tallies() {
+  std::list<pyne::Tally> :: iterator it;
+
+  std::string tally_destination = "/tally"; // to avoid compiler warning
+
+  // loop over the processed material library and write each one to the file
+  for (it = uwuw_tally_library.begin() ; it != uwuw_tally_library.end() ; ++it) {
+    // the current tally
+    pyne::Tally tally = *it;
+    if (verbose) {
+      std::cout << "Writing tally " << tally.tally_name;
+      std::cout << " to file " << output_filename << std::endl;
+    }
+
+    // write the material to the file
+    tally.write_hdf5(output_filename, tally_destination);
+  }
+
+  return;
+}
+
+
+// creates a new material using the Material material as a basis
+pyne::Material uwuw_preprocessor::create_new_material(pyne::Material material, std::string density) {
+  pyne::Material new_mat; // to return
+  pyne::comp_map comp = material.comp;
+
+  // make sure to bring metadata with us
+  new_mat = pyne::Material(comp, 1.0,
+                           material.density,
+                           material.atoms_per_molecule,
+                           material.metadata);
+
+  // use the name concatenator to make the fluka name
+  std::string fluka_name = ncr->make_name_8bytes(material.metadata["name"].asString());
+
+  std::string material_name;
+  // make a new name
+  if (density != "") {
+    material_name = "mat:" + material.metadata["name"].asString() + "/rho:" + density;
+    new_mat.density = atof(density.c_str());
+  } else {
+    material_name = "mat:" + material.metadata["name"].asString();
```

Comment:
This if-else could be simplified a little by doing the following part only once:
```c++
    material_name = "mat:" + material.metadata["name"].asString();
```